### PR TITLE
Feature/OP-1549: NYC.ID v2.0

### PR DESCRIPTION
**Pull request by joelbcastillo**
You will need to follow the steps outlined in docs/saml_nycid_setup.md to setup local NYC.ID authentication using the non-production environment.

To test this locally you will also need to perform the following steps:
1) Copy a production data set to your local database:
```bash
sudo service rh-postgresql95-postgresql restart
sudo -u postgres /opt/rh/rh-postgresql95/root/usr/bin/dropdb openrecords
sudo -u postgres /opt/rh/rh-postgresql95/root/usr/bin/createdb openrecords
sudo -u postgres /opt/rh/rh-postgresql95/root/usr/bin/pg_restore -U developer -h 127.0.0.1 -d openrecords < /path/to/your/sql.pgdump
```
2) Due to a migration issue in 2018, there are a number of non-existent entries in the UserRequests table that need to be removed:
```bash
psql -U developer -h 127.0.0.1 -d openrecords -c "DELETE FROM user_requests WHERE NOT EXISTS(SELECT guid FROM users WHERE users.guid = user_requests.user_guid)"
```
3) The upgrade process needs to be run in 4 steps:
a) Run the first upgrade script: `python manage.py db upgrade 89d3f2e347f1`
b)Uncomment lines 339 - 350 in `models.py` to add auth_user_type into the Users table only
c) Run the custom migration script to update the users and keep their current designations: `python manage.py migrate_nycid_users`
d) Run the final upgrade script: `python manage.py db upgrade`
e) Comment out / remove lines 339-350 from `models.py`
4) Since we are dealing with production data dump, make sure to remove any real email addresses.
```bash
psql -U developer -h 127.0.0.1 -d openrecords -c "UPDATE users SET notification_email = 'test@mailinator.com'; UPDATE agencies SET default_email = 'test@mailinator.com'; UPDATE agencies SET appeals_email = 'test@mailinator.com';"
```
5) Update your ElasticSearch index: `python manage.py es_recreate`

**NOTE**: When using a production database dump, you will need at least 8GB of memory for your Vagrant machine to allow ElasticSearch to create an index of the database.

### CODE REVIEW

**Comment by joelbcastillo** [comment]
This pull request **introduces 5 alerts** and **fixes 3** when merging 99a34ef90a680f606f75af7eb37066f89023af51 into 7e9472d482a71496e0aa4ed9fc7e609df4ce059f - [view on LGTM.com](https://lgtm.com/projects/g/CityOfNewYork/NYCOpenRecords/rev/pr-df73170c893fe629beccaec4880f98f6b8006e62)

**new alerts:**

* 2 for Syntax error
* 1 for Unused local variable
* 1 for Except block handles &#39;BaseException&#39;
* 1 for Variable defined multiple times

**fixed alerts:**

* 2 for Unreachable code
* 1 for Unused import

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by joelbcastillo** [comment]
This pull request **introduces 5 alerts** and **fixes 3** when merging a5a43c249f8d2f7ee8255f99ee2f39975cbfb656 into 7e9472d482a71496e0aa4ed9fc7e609df4ce059f - [view on LGTM.com](https://lgtm.com/projects/g/CityOfNewYork/NYCOpenRecords/rev/pr-028df58845934b1e8d41767432bc17cf6f239b63)

**new alerts:**

* 2 for Syntax error
* 1 for Unused local variable
* 1 for Except block handles &#39;BaseException&#39;
* 1 for Variable defined multiple times

**fixed alerts:**

* 2 for Unreachable code
* 1 for Unused import

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by joelbcastillo** [comment]
This pull request **introduces 5 alerts** and **fixes 3** when merging 7115626ef84ca8f4cde5c7db3bf480bb27c4b59c into 7e9472d482a71496e0aa4ed9fc7e609df4ce059f - [view on LGTM.com](https://lgtm.com/projects/g/CityOfNewYork/NYCOpenRecords/rev/pr-2bca94fe40a33d624ce89256e8fa538c5bfe6454)

**new alerts:**

* 2 for Syntax error
* 1 for Unused local variable
* 1 for Except block handles &#39;BaseException&#39;
* 1 for Variable defined multiple times

**fixed alerts:**

* 2 for Unreachable code
* 1 for Unused import

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by joelbcastillo** [comment]
This pull request **introduces 5 alerts** and **fixes 3** when merging 2e5b6d8d392933ab8c1f551f215e84245f6e2173 into 7e9472d482a71496e0aa4ed9fc7e609df4ce059f - [view on LGTM.com](https://lgtm.com/projects/g/CityOfNewYork/NYCOpenRecords/rev/pr-fc6f22dc8aa295843376c5930dbbd463b5f34d37)

**new alerts:**

* 2 for Syntax error
* 1 for Unused local variable
* 1 for Except block handles &#39;BaseException&#39;
* 1 for Variable defined multiple times

**fixed alerts:**

* 2 for Unreachable code
* 1 for Unused import

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by joelbcastillo** [comment]
This pull request **introduces 12 alerts** and **fixes 11** when merging ae1e5a8a6febb5979d51bc05174c281b385dd1a7 into 46be668ee2a6c39a57e781ec5938fa0ac1fbfadb - [view on LGTM.com](https://lgtm.com/projects/g/CityOfNewYork/NYCOpenRecords/rev/pr-874ad6d494dc53149f3de8e097575336ba0c78b7)

**new alerts:**

* 3 for Unused local variable
* 3 for Except block handles &#39;BaseException&#39;
* 2 for Testing equality to None
* 2 for \_\_init\_\_ method calls overridden method
* 1 for Unused named argument in formatting call
* 1 for Variable defined multiple times

**fixed alerts:**

* 6 for Syntax error
* 3 for Unreachable code
* 1 for Unused import
* 1 for List comprehension variable used in enclosing scope

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: app/user/views.py [hunk 1]
```diff
@@ -63,303 +53,312 @@ def patch(user_id):
...
+        if not agency and agency.is_active:
+            return jsonify({'error': 'Agency must exist in the database and be active'}), 400
+
+        if not current_user.is_super:
+            # Current user must belong to agency specified by agency_ein
+            current_user_is_agency_admin = current_user.is_agency_admin(agency.ein)
+
+            # user_in_agency = agency_ein in [agency.ein for agency in user_.agencies.all()]
+            user_in_agency = AgencyUsers.query.filter(AgencyUsers.user_guid == user_.guid, AgencyUsers.agency_ein == agency_ein).one_or_none()
+            if not current_user_is_agency_admin:
```
Comment (gzhou1) [inline app/user/views.py]: Maybe this should go before line 104?

File: app/user/views.py [hunk 2]
```diff
@@ -63,303 +53,312 @@ def patch(user_id):
...
+        # Agency must exist and be active to modify users
+        agency = Agencies.query.filter_by(ein=agency_ein).one_or_none()
+        if not agency and agency.is_active:
+            return jsonify({'error': 'Agency must exist in the database and be active'}), 400
+
+        if not current_user.is_super:
+            # Current user must belong to agency specified by agency_ein
+            current_user_is_agency_admin = current_user.is_agency_admin(agency.ein)
+
+            # user_in_agency = agency_ein in [agency.ein for agency in user_.agencies.all()]
```
Comment (gzhou1) [inline app/user/views.py]: Remove if not needed.

**Comment by joelbcastillo** [comment]
This pull request **introduces 9 alerts** and **fixes 11** when merging 808018ac201bcd527a4362b9fbf1a3a79f5d75bc into 46be668ee2a6c39a57e781ec5938fa0ac1fbfadb - [view on LGTM.com](https://lgtm.com/projects/g/CityOfNewYork/NYCOpenRecords/rev/pr-b8626d1582ad0eab93ca67d65a53b6e4503922d3)

**new alerts:**

* 3 for Unused local variable
* 2 for Testing equality to None
* 2 for \_\_init\_\_ method calls overridden method
* 1 for Except block handles &#39;BaseException&#39;
* 1 for Variable defined multiple times

**fixed alerts:**

* 6 for Syntax error
* 3 for Unreachable code
* 1 for Unused import
* 1 for List comprehension variable used in enclosing scope

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: app/response/utils.py [hunk 3]
```diff
@@ -1785,8 +1780,7 @@ def _file_email_handler(request_id, data, page, agency_name, email_template):
                                                 page=page,
                                                 agency_name=agency_name,
                                                 agency_default_email=request.agency.default_email,
-                                                public_requester=request.requester.auth_user_type in
-                                                                 user_type_auth.PUBLIC_USER_TYPES,
+                                                public_requester=(request.requester.is_nyc_employee is False),
```
[the remark this answers is not in the document]
Comment (joelbcastillo) [reply app/response/utils.py]: @gzhou1 I think this is fixed in the latest version.

**Comment by joelbcastillo** [comment]
This pull request **introduces 9 alerts** and **fixes 11** when merging 1e48f7a6d46439a33a1c1464f7167d6256ba5d0c into 46be668ee2a6c39a57e781ec5938fa0ac1fbfadb - [view on LGTM.com](https://lgtm.com/projects/g/CityOfNewYork/NYCOpenRecords/rev/pr-15fee0bd6905bfe290162d8e0a3fa1a93af49e76)

**new alerts:**

* 3 for Unused local variable
* 2 for Testing equality to None
* 2 for \_\_init\_\_ method calls overridden method
* 1 for Except block handles &#39;BaseException&#39;
* 1 for Variable defined multiple times

**fixed alerts:**

* 6 for Syntax error
* 3 for Unreachable code
* 1 for Unused import
* 1 for List comprehension variable used in enclosing scope

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by joelbcastillo** [comment]
This pull request **introduces 9 alerts** and **fixes 11** when merging 12e2275a50142fb4e253bfa84ee502405454a3b0 into 4ea3587becd02f9371e2d209da79f2b3dae274ad - [view on LGTM.com](https://lgtm.com/projects/g/CityOfNewYork/NYCOpenRecords/rev/pr-5fabcb159af11961ab7a54c363d191e5be903e25)

**new alerts:**

* 3 for Unused local variable
* 2 for Testing equality to None
* 2 for \_\_init\_\_ method calls overridden method
* 1 for Except block handles &#39;BaseException&#39;
* 1 for Variable defined multiple times

**fixed alerts:**

* 6 for Syntax error
* 3 for Unreachable code
* 1 for Unused import
* 1 for List comprehension variable used in enclosing scope

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by joelbcastillo** [comment]
This pull request **introduces 12 alerts** and **fixes 11** when merging 9a67e826878d0f8426ed56c150aab6ae8275a539 into 4ea3587becd02f9371e2d209da79f2b3dae274ad - [view on LGTM.com](https://lgtm.com/projects/g/CityOfNewYork/NYCOpenRecords/rev/pr-14a404953445902ce5dc93a9d5c7b142cf0127a7)

**new alerts:**

* 3 for Unused local variable
* 3 for Unused import
* 2 for Testing equality to None
* 2 for \_\_init\_\_ method calls overridden method
* 1 for Except block handles &#39;BaseException&#39;
* 1 for Variable defined multiple times

**fixed alerts:**

* 6 for Syntax error
* 3 for Unreachable code
* 1 for Unused import
* 1 for List comprehension variable used in enclosing scope

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Pull request merged**